core/src/main/java/io/seata/core/rpc/netty/NettyClientChannelManager.java

```diff
@@ -175,38 +172,95 @@ void reconnect(String transactionServiceGroup) {
         }
     }
     
-    void invalidateObject(final String serverAddress, final Channel channel) throws Exception {
-        nettyClientKeyPool.invalidateObject(poolKeyMap.get(serverAddress), channel);
-    }
-    
-    void registerChannel(final String serverAddress, final Channel channel) {
-        if (null != channels.get(serverAddress) && channels.get(serverAddress).isActive()) {
-            return;
-        }
-        channels.put(serverAddress, channel);
-    }
-    
     private Channel doConnect(String serverAddress) {
         Channel channelToServer = channels.get(serverAddress);
         if (channelToServer != null && channelToServer.isActive()) {
             return channelToServer;
         }
         Channel channelFromPool;
+        NettyPoolKey poolKey;
         try {
             NettyPoolKey currentPoolKey = poolKeyFunction.apply(serverAddress);
             NettyPoolKey previousPoolKey = poolKeyMap.putIfAbsent(serverAddress, currentPoolKey);
             if (null != previousPoolKey && previousPoolKey.getMessage() instanceof RegisterRMRequest) {
                 RegisterRMRequest registerRMRequest = (RegisterRMRequest) currentPoolKey.getMessage();
                 ((RegisterRMRequest) previousPoolKey.getMessage()).setResourceIds(registerRMRequest.getResourceIds());
             }
-            channelFromPool = nettyClientKeyPool.borrowObject(poolKeyMap.get(serverAddress));
+            poolKey = previousPoolKey != null ? previousPoolKey : currentPoolKey;
```

Comment:
previousPoolKey already judgment in if condtion.

Reply:
previousPoolKey could still be null.

Reply:
it's my fault.